### DISCUSSION TITLE
Linux调用Alist进行上传、下载和删除文件操作命令分享

### DISCUSSION THREAD

**Discussion by qq2267222680** [AlistGo/alist#2061] [Show and tell]
## variables
```shell
file='google-research.tar'
domain='http://alist.xxx.com'
token='xxxxxx' # v2.6.1的token在前端页面登录后的的请求头中获取，v3.2.0的token在前端页面登录后从登录响应中拿到。
```

## usage v2.6.1
```shell
curl -# -X POST "$domain/api/public/upload" -F "path=/tourist" -F "files=@$file" | tee /dev/null && echo ''
curl -# -O "$domain/d/tourist/$file"
curl -X DELETE "$domain/api/admin/files" -d "{\"names\":[\"$file\"],\"path\":\"/tourist\"}" -H 'Authorization: $token' -H 'Content-Type: application/json;charset=UTF-8' && echo ''
```

## usage v3.2.0
```shell
curl -# -T "$file" "$domain/api/fs/put" -H "File-Path: /tourist/$(basename $file)" | tee /dev/null && echo ''
curl -# -O "$domain/d/tourist/$(basename $file)" # wget "$domain/d/tourist/$(basename $file)"
curl -X POST "$domain/api/fs/remove" -d "{\"names\":[\"$(basename $file)\"],\"dir\":\"/tourist\"}" -H 'Content-Type: application/json;charset=UTF-8' \
  -H "Authorization: $token" \
  && echo ''
```

**Comment by qq2267222680** [comment]
竟然没人回复，记得回复一下呀兄弟们，让该分享保持热度呀

**Reply by anwen-anyi** [reply]
东西是好东西...会用的会用，不会用的看了也不会 哈哈

**Reply by qq2267222680** [reply]
现如今工作中对rzsz支持的工具也渐渐出现各种BUG，所以才想到用Alist进行文件中转。主要是这些用法官方也没提供，这两个版本我都各自调了好几个小时才找到合适的命令的，也不想浪费，多分享给一个人就能增加它的一分价值。

**Comment by naicfeng** [comment]
测试你代码遇到上传失败
```
(*MediaTrack).request() {"status":"FAILURE","status_code":422,"code":10002,"reason":"UNKNOWN_REASON","message":"参数不合法","errors":[{"resource":"createAssetRequest","filed":"Mime","error":"Mime can not be empty when asset type is file"}],"path":"/v3/assets/*/children","trace_id":"*","requestId":""} 
```
某些 `Drive` (如MediaTrack)上传需要指定 `mime` ，流上传模式是解析本地提供的`content-type`头，你这没提供上传会失败。
当然在浏览器端上传是没问题的，会提供`content-type`
https://github.com/alist-org/alist/blob/5f79d665d9ae3a9b70365ae1078c4754b987f225/server/handles/fsup.go#L56

Form模式上传会自动解析
https://github.com/alist-org/alist/blob/5f79d665d9ae3a9b70365ae1078c4754b987f225/server/handles/fsup.go#L121

所以还是用Form模式上传比较稳妥

```
curl -X 'PUT' "$domain/api/fs/form" -H "File-Path: /tourist/$(basename $file)" -H "Authorization: $token" -F file="@$file"
```

**Comment by GengGode** [comment]
厉害，就在找这个东西，等下测试看看

**Comment by vincentTea** [comment]
有没有指定文件生成短链的

**Comment by Milky-Liu** [comment]
完了……小白看不懂T_T

**Comment by GengGode** [comment]
返回413 Request Entity Too Large有什么解决方案吗，1MB以下的都能传，大点的就不行了

**Reply by cloudswave** [reply]
可以尝试使用rclone工具
```
screen rclone move -v /local/path alist-webdav:/webdav/path -P --retries 200 # 显示进度 失败重试200次
```

**Reply by templatejacob** [reply]
你是不是配置了Nginx，Nginx中有request body的限制在server {}里面配置client_max_body_size 90000m;即可，大小自己来定。

**Reply by GengGode** [reply]
确实，感谢，设置了就好了

**Reply by cddone** [reply]
超过1G的设置了还是报413

**Comment by cloudswave** [comment]
我尝试使用davfs挂载但是如果driver服务商速度不行cp大文件容易导致失败，目前使用rclone来上传到webdav，把重试次数增大，还可以保证大文件能上传成功

**Comment by CollinChan002** [comment]
好帖！谢谢！帮了我大忙

**Comment by youtehub** [comment]
来顶贴一下，保持热度，难得指令实现

**Comment by sangemajia** [comment]
虽然看不懂 不过好厉害

**Comment by sangemajia** [comment]
# 使用脚本上传文件到网盘。先登录alist挂载网盘

file='上传文件名(带文件路径)'
name='网盘名称'
domain='http://域名或IP:5244'
token='alistv3版的token在登录alist后打开Edge浏览器(F12)-开发者工具-网络-名称（get或者list）找Authorization:后面就是token'
curl -X 'PUT' "$domain/api/fs/form" -H "File-Path: /$name/$(basename $file)" -H "Authorization: $token" -F file="@$file"


这里用的是@naicfeng的代码 游客不能上传

**Comment by nibazshab** [comment]
游客可以使用如下指令上传，不需要带 token，其中 `File-Path: /` 指向在为 guest 用户设置的基本路径

```
curl -X 'PUT' "域名/api/fs/form" -H "File-Path: /文件名" -F file="@文件名"
```

据我所知，很多第三方临时网盘服务网站，都可以 `curl -F file=@文件名 域名` 直接上传，示例： `curl -F file=@文件名 file.io`

而目前没有找到为 alist 省略掉 `-H "File-Path: /文件名"` 以及 `-X 'PUT'` 的方法

**Reply by nibazshab** [reply]
使用 WebDAV 方式上传，而不使用 api 的命令如下。如果想以游客身份上传，需要在后台为游客打开 WebDAV 权限，游客的用户名和密码均为 __guest__

```sh
curl -u 用户名:密码 -T 文件 域名/dav/
```

示例：`curl -u guest:guest -T 文件 域名/dav/`

**Comment by ghost** [comment]
今天试了好久才成功的命令,大家可以试试
name='文件'
domain='http://xxxxx:xx'
token='alist-xxxxxx-xxxxxxxxx-xxxxxxxxxxxxxxx'
#上传(会自动创建子目录)
curl -X PUT "$domain/api/fs/put" -H "Authorization:$token" -H "File-Path:/HMHY/test/11/$name" -T "/root/$name"
#删除
curl -X POST "$domain/api/fs/remove" -H "Authorization: $token" -d "{\"names\":[\"$name\"],\"dir\":\"/HMHY/test\"}" -H 'Content-Type: application/json;charset=UTF-8'
#创建文件夹
curl -X POST "$domain/api/fs/mkdir" -H "Authorization: $token" -d "{\"path\":\"/HMHY/test/$name\"}" -H 'Content-Type: application/json;charset=UTF-8'
#重命名
curl -X POST "$domain/api/fs/rename" -H "Authorization: $token" -d "{\"name\":\"$name1\",\"path\":\"/HMHY/test/$name\"}" -H 'Content-Type: application/json;charset=UTF-8'

**Comment by pongfcnkl** [comment]
如何用curl调用复制命令？
![QQ截图20230728224220](https://github.com/alist-org/alist/assets/24273319/79115da3-d026-40a3-b452-922469c636d1)

**Comment by yr-github** [comment]
有没有docker适用的命令分享呀？我的这个是在路由器的docker，路由器没开ssh，我想实现后台下载到硬盘上。

**Reply by qq2267222680** [reply]
Mac的iterm2可以使用it2dl命令
其他情况可以使用通用的lrzsz命令。

**Comment by SkysCrystal** [comment]
请问这个token多久有效？

**Reply by ifloppy** [reply]
Token有效期有限的。我自己之前用的解决办法是再加一个用curl发请求登录，再把token传进去用

**Reply by flashlab** [reply]
token=$(curl -s '$domain/api/auth/login/hash' \
  -H 'Content-Type: application/json;charset=UTF-8' \
  --data-raw '{"username":"$username","password":"$pwdhash","otp_code":""}' | jq .data.token -r)

**Comment by KnIfER** [comment]
好麻烦，就不能 `alist.exe 上传.rar .../到哪儿/`

**Comment by KnIfER** [comment]
好麻烦，就不能 `alist.exe 上传.rar .../到哪儿/`

**Reply by pongfcnkl** [reply]
rclone可以

**Comment by Cyang39** [comment]
感谢，很好用！编写了个 shell 脚本，简化了从 token 获取到上传文件的流程：[Gist](https://gist.github.com/Cyang39/8c9de554178302ecc5626b0698c0fcfa)
